L1Trigger/L1THGCal/src/concentrator/HGCalConcentratorSuperTriggerCellImpl.cc

```diff
@@ -86,53 +131,169 @@ HGCalConcentratorSuperTriggerCellImpl::getSuperTriggerCellId(int detid) const {
 
       }
 
-      TC_12th =  (rocnum << kRocShift_) | ( (Uprime << kUShift_ | Vprime) & kSplit_v9_ );
+      int TC_split = -1;
 
-      int TC_split =  TC_12th;
-      if (stcSize_.at(thickness) == kSTCsizeCoarse_){
-        TC_split =  rocnum;
+      if ( STCsize > 0 ){
+        TC_split =  (rocnum << kRocShift_) | ( (Uprime << kUShift_ | Vprime) & kSplit_v9_.at( 2 ) );
       }
+      else{
+        
+        if (stcSize_.at(thickness) == kSTCsizeCoarse_){
+          TC_split =  rocnum;
+        }
+        else{
+          TC_split =  (rocnum << kRocShift_) | ( (Uprime << kUShift_ | Vprime) & kSplit_v9_.at( stcSize_.at(thickness) ) );
+        }
 
+      }
+    
       return TC_wafer<<kWafer_offset_ | TC_split;
       
     }
     
     
-
   }
   else{
     return -1;
   }
   
 }
 
+void 
+HGCalConcentratorSuperTriggerCellImpl::
+createMissingTriggerCells( std::unordered_map<unsigned,SuperTriggerCell>& STCs, std::vector<l1t::HGCalTriggerCell>& trigCellVecOutput) const
+{ 
+
+    for (const auto& s: STCs){
+
+      //      bool satisfy = false;
+      //Find and create missing TCs (for super TC 4)
+      int thickness = triggerTools_.thicknessIndex(s.second.GetSTCId(),true);
+      
+      int kSplitInv = ~( (~kSixBitMax_) | kSplit_.at ( stcSize_.at(thickness) ) );
+      int tc_base = s.second.GetSTCId();
+
+      for ( int i = 0; i < kSplitInv + 1 ; i++ ){
+        if (  (i & kSplitInv)!=i         )  continue; 
+
+
+	if ( fixedDataSize_ ){
+	  if ( thickness > 0 ){
+	    if (i&1) continue;
+	  }
+	}
+        
+        int newtc = tc_base | i;
+        
+        if ( !triggerTools_.validTriggerCell(newtc) ) {
+          continue;
+        }
+        
+        l1t::HGCalTriggerCell triggerCell;
+        GlobalPoint point = triggerTools_.getTCPosition(newtc);
+        math::PtEtaPhiMLorentzVector p4(0, point.eta(), point.phi(), 0.);
+        triggerCell.setPosition(point);
+        triggerCell.setP4(p4);
+        triggerCell.setDetId(newtc);
+        
+        trigCellVecOutput.push_back ( triggerCell );
+        
+      }
+    }
+}
+
+
+void 
+HGCalConcentratorSuperTriggerCellImpl::
+coarsenTriggerCells(  std::vector<l1t::HGCalTriggerCell>& trigCellVecInput )
+{ 
+
+  for  (l1t::HGCalTriggerCell & tc : trigCellVecInput) {
+     if (tc.subdetId() == HGCHEB) continue;
+
+     int thickness = triggerTools_.thicknessIndex(tc.detId(),true);
+     if ( thickness > 0 ) continue;
+
+     int detid = tc.detId();
+     
+     if ( detid & 1) { //if even
+       std::cout << "NOT EVEN ? " << std::endl;
+     }
+   
+     int detid_odd = detid | 1;
+     GlobalPoint point_even = triggerTools_.getTCPosition(detid);
+     GlobalPoint point_odd = triggerTools_.getTCPosition(detid_odd);
+     GlobalPoint average_point  ( (point_even.x()+point_odd.x())/2,(point_even.y()+point_odd.y())/2,(point_even.z()+point_odd.z())/2  );     
```

Comment:
The computation of Coarse TC position should be externalized from the the SuperTC implementation.